core/desugarMutual.ml

```diff
@@ -0,0 +1,152 @@
+open Sugartypes
+open Utility
+open SourceCode
+open Errors
+
+(* Desugars Mutual. *)
+(*
+ * mutual {
+ *   typename Foo = ...;
+ *   typename Bar = ...;
+ *   fun foo() {
+ *       ...
+ *   }
+ *
+ *  fun bar() {
+ *       ...
+ *   }
+ * } -->
+ * Typenames [Foo; Bar]
+ * Funs [foo; bar]
+ *
+ * Any other binding occurring in a mutual block is forbidden.
+ *)
+
+(* Simple check to see whether a function is recursive: checks whether the
+ * binder occurs within the free variables of the function body. *)
+let is_recursive bnd fnlit =
+  StringSet.mem (Binder.to_name bnd) (Freevars.funlit fnlit)
+
+
+(* Fun bindings must be lifted into `Funs if they are recursive. This is
+ * performed after the first pass. *)
+let lift_funs =
+object ((self : 'self_type))
+    inherit SugarTraversals.map as super
+    method! binding = fun b ->
+      let position = WithPos.pos b in
+      match WithPos.node b with
+      |  Fun (bndr, lin, (tvs, fnlit), location, dt) ->
+          let fnlit = self#funlit fnlit in
+          if is_recursive bndr fnlit then
+            WithPos.make ~pos:position (Funs [(bndr, lin, ((tvs, None), fnlit), location, dt, position)])
+          else
+            WithPos.make ~pos:position (Fun (bndr, lin, (tvs, fnlit), location, dt))
+      | _ -> super#binding b
+end
+
+let check_dups funs tys =
+  (* Check to see whether there are any duplicate names, and report
+   * an error if so. *)
+  let assoc_to_list_map get_name xs =
+    List.fold_left (fun acc (x, pos) ->
+      let name = get_name x in
+      StringMap.update name (fun x_opt ->
+        OptionUtils.opt_app
+          (fun positions -> Some (pos :: positions))
+          (Some [pos]) x_opt) acc) StringMap.empty xs in
+
+  let funs_map =
+    assoc_to_list_map
+      (fun (bndr, _, _, _, _) -> Binder.to_name bndr) funs in
+  let tys_map = assoc_to_list_map (fst3) tys in
+  let check map =
+    let dups =
+      StringMap.filter (fun _ poss -> List.length poss > 1) map in
+    if StringMap.cardinal dups > 0 then
+      raise (MultiplyDefinedMutualNames dups)
+    else () in
+  check funs_map; check tys_map
+
+
+let rec flatten_simple = fun () ->
+object(self)
```

Comment:
Why not 
```
let rec flatten_simple () = 
...
and flatten_bindings () = 
```

??

Also, I'd prefer if the body of both "let"s were indented, so that the `open` and `object ... end` constructs are clearly not toplevel ones.